generators/csharp/dynamic-snippets/src/context/DynamicTypeMapper.ts

```diff
@@ -36,13 +38,16 @@ export class DynamicTypeMapper {
                 return this.convertNamed({ named });
             }
             case "optional":
-                return csharp.Type.optional(this.convert({ typeReference: args.typeReference.value }));
-            case "nullable":
-                return csharp.Type.optional(this.convert({ typeReference: args.typeReference.value }));
+            case "nullable": {
+                const value = this.convert({ typeReference: args.typeReference.value });
+                return args.unboxOptionals
+                    ? value
+                    : csharp.Type.optional(value);
+            }
             case "primitive":
                 return this.convertPrimitive({ primitive: args.typeReference.value });
             case "set":
-                return csharp.Type.set(this.convert({ typeReference: args.typeReference }));
+                return csharp.Type.set(this.convert({ typeReference: args.typeReference, unboxOptionals: true }));
```

Comment:
Why unbox optionals here?

Reply:
Same as the reason outlined [here](https://github.com/fern-api/fern/pull/6326#discussion_r1985475599).